wordpress-mobile/WordPressKit-iOS
# Refactor retrieving org REST API nonce

### Description

> [!Note]
> This PR is built on top of https://github.com/wordpress-mobile/WordPressKit-iOS/pull/666. Also, it might be easier to review commit-by-commit.

Currently the retrieving nonce code is coupled with `CookieNonceAuthenticator` and Alamofire. This PR moves them out as of the authenticator implementation and makes it depending on `URLSession`. The new code will be used in upcoming .org REST API client refactor.

#### Authentication Details

.org REST API uses [cookie+nonce authentication](https://developer.wordpress.org/rest-api/using-the-rest-api/authentication/#cookie-authentication). After user logins into the app, the app _should_ have the authentication related cookies available, which will be added automatically by `URLSession`. The `nonce` part is what we need to active find.

The library finds a nonce from either `wp-admin/post-new.php` webpage html or `wp-admin/admin-ajax.php?action=rest-nonce`. Both webpages require user signed in (authenticated via cookies, like all webpages). If user has not signed in, both webpages return non-200 status code (which can be tested using `curl -I https://<self-host-site>/<link-to-webpage>`). In that case, the library performs a sign in HTTP request with a redirection to one of the above two webpages. 

The nonce will be passed to .org REST API via `X-WP-Nonce` header. With cookies and nonce available to REST API, user can now make authenticated API calls.

#### A small change in this PR

At the moment, the library always makes the aforementioned sign in requests, which I think in most cases are redundant. Because at the time a org REST API is called, user should already signed in to the app. So the small change I made in this PR was, always getting the webpage content as the first step, and signing in with a redirection link as the fallback.

### Testing Details

I have added a few unit tests in this library. `NonceRetrievalTests.swift` is pretty much a duplication of `CookieNonceAuthenticatorTests` (which will be removed along with Alamofire).

To test this in a real app:
1. Update the Podfile in your local WordPress-iOS checkout to depend on this PR branch.
2. Set a breakpoint at [this line](https://github.com/wordpress-mobile/WordPressKit-iOS/blob/76e18a702b010b2c53563fb5d3906e107948e874/WordPressKit/WordPressOrgRestApi.swift#L64).
3. Launch the app from Xcode.
4. Sign in using a self hosted site
5. Go to Posts and open an existing post (Gutenberg editor makes .org REST API requests).
6. In a couple seconds, Xcode should stop at the breakpoint set above. Observe the API result: it should succeed.

---
